### PR TITLE
[mapreduce-tidb-connector] mapreduce read from tikv

### PR DESCRIPTION
**Pull request by mengxin9014**
support TiDBInputFormat for mapreduce to read from Tikv

### CODE REVIEW

File: mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2020 TiDB Project Authors.
```
Comment (marsishandsome) [inline mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBConfiguration.java:2]:
```suggestion
 * Copyright 2021 TiDB Project Authors.
```

File: mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBInputFormat.java [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    this.tableHandleInternal = new TableHandleInternal(
+        UUID.randomUUID().toString(), databaseName, tableName);
+    List<ColumnHandleInternal> columns = clientSession.getTableColumns(tableHandleInternal)
+        .orElseThrow(() -> new NullPointerException("columnHandleInternals is null"));
+    Map<String, Integer> nameAndIndex = new HashMap<>();
+
+    IntStream.range(0, columns.size())
+        .forEach(i -> nameAndIndex.put(columns.get(i).getName(), i));
+
+    String[] fieldNames =
```
Comment (marsishandsome) [inline mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBInputFormat.java:121]: what if `fieldNames.length = 0` or ` fieldNames = null` ?

File: mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBConfiguration.java (hunk 1 above)
Comment (mengxin9014) [reply mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBConfiguration.java]: done

File: mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBInputFormat.java (hunk 2 above)
Comment (mengxin9014) [reply mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBInputFormat.java:121]: it‘s going to be judged by the setInput method in class TiDBInputFormat.

File: mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/example/TiDBMapreduceDemo.java [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    protected void setup(Context context) {
+      System.out.println("job attempt ID : " + context.getTaskAttemptID());
+      printFields(context.getConfiguration().getStrings("tidb.field.names"));
+    }
+
+    @Override
+    protected void map(LongWritable key, TiDBRowData value, Context context) {
+      // do mapper here
+      printRowData(key.get(), value.toString());
+
```
Comment (marsishandsome) [inline mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/example/TiDBMapreduceDemo.java:71]: delete this line?

File: mapreduce/mapreduce-base/src/main/test/java/io/tidb/bigdata/mapreduce/tidb/MapReduceTest.java [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+      Assert.assertEquals("2020-08-10 16:30:29.0", c24.toString());
+      Assert.assertEquals(new Integer(2020), c25);
+      Assert.assertTrue(c26);
+      Assert.assertEquals("{\"a\":1,\"b\":2}", c27);
+      Assert.assertEquals("1", c28);
+      Assert.assertEquals("a", c29);
+    }
+
+    clientSession.close();
+
```
Comment (marsishandsome) [inline mapreduce/mapreduce-base/src/main/test/java/io/tidb/bigdata/mapreduce/tidb/MapReduceTest.java:285]: delete this line?

File: mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/MapreduceCmd.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
+/*
```
Comment (marsishandsome) [inline mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/MapreduceCmd.java:1]: move MapreduceCmd  class to example?

File: mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/MapreduceCmd.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+  public String timestamp;
+
+  @Parameter(names = {"-limit", "-l"}, description = "record limit per mapper")
+  public Integer limit = Integer.MAX_VALUE;
+
+  public MapreduceCmd(String[] args) {
+    JCommander.newBuilder().addObject(this).build().parse(args);
+  }
+
+  public Configuration toConf() {
```
Comment (marsishandsome) [inline mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/MapreduceCmd.java:59]: this function is never be used?

File: mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+  public static void configureDB(Configuration conf,
+      String dbUrl, String databaseName, String userName, String password) {
+
+    conf.set(URL_PROPERTY, dbUrl);
+    conf.set(DATABASE_NAME, databaseName);
+    conf.set(USERNAME_PROPERTY, userName);
+    conf.set(PASSWORD_PROPERTY, password);
+  }
+
+  private Configuration conf;
```
Comment (marsishandsome) [inline mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBConfiguration.java:81]: final ?

File: mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+
+  public TiDBConfiguration(Configuration conf) {
+    this.conf = conf;
+  }
+
+  public ClientSession getSingleConnection() {
+
+    Map<String, String> properties = new HashMap<>();
```
Comment (marsishandsome) [inline mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBConfiguration.java:89]:
```suggestion
    Map<String, String> properties = new HashMap<>(3);
```

File: mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBInputSplit.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+public class TiDBInputSplit extends InputSplit implements Writable {
+  private String startKey;
+  private String endKey;
+  private String connectorId;
+  private String schemaName;
+  private String tableName;
+
+  public static final String [] EMPTY_STRING_ARRAY = new String[0];
+
+  public TiDBInputSplit(){
```
Comment (marsishandsome) [inline mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBInputSplit.java:34]:
```suggestion
  public TiDBInputSplit() {
```

File: mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBInputSplit.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+  @Override
+  public void readFields(DataInput in) throws IOException {
+    this.startKey = in.readUTF();
+    this.endKey = in.readUTF();
+    this.connectorId = in.readUTF();
+    this.schemaName = in.readUTF();
+    this.tableName = in.readUTF();
+  }
+
```
Comment (marsishandsome) [inline mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBInputSplit.java:94]: delete this line

File: mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBResultSet.java [hunk 11]
```diff
@@ -0,0 +1,1069 @@
...
+  }
+
+  @Override
+  public boolean rowDeleted() throws SQLException {
+    throw new IllegalArgumentException("no function");
+  }
+
+  @Override
+  public void updateBoolean(int columnIndex, boolean x) throws SQLException {
+
```
Comment (marsishandsome) [inline mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBResultSet.java:441]: throw error?

File: mapreduce/README.md [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+cd TiBigData
+mvn clean package -DskipTests -am -pl mapreduce/mapreduce-base
+cp mapreduce/mapreduce-2.0/target/mapreduce-tidb-connector-base-0.0.4-SNAPSHOT.jar ${HOME}/lib
+```
+
+## Version
+
+`hadoop 2.X` and `hadoop 3.X` are supported.
+
+
```
Comment (marsishandsome) [inline mapreduce/README.md:16]: delete this line

File: mapreduce/README.md [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+hadoop jar mapreduce-tidb-connector-base-0.0.4-SNAPSHOT.jar io.tidb.bigdata.mapreduce.tidb.examples.TiDBMapreduceDemo  \
+ -du  ${DATABASE_URL} \
+ -u   ${USERNAME} \
+ -p   ${PASSWORD}  \
+ -dn  ${TIDB_DATABASE} \
+ -n   ${TABLE_NAME} \
+ -f   ${FIELD1} \
+ -f   ${FIELD2} \
+ -t   ${TABLE_NAME}
+ -ts  ${SNAPSHOT_TIMESTAMP} \
```
Comment (marsishandsome) [inline mapreduce/README.md:29]: users do not know the format of `SNAPSHOT_TIMESTAMP`

File: mapreduce/README.md [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+ -l   ${LIMIT} 
+```
+### example
+
```
Comment (marsishandsome) [inline mapreduce/README.md:33]: Step 1: create table ...

File: mapreduce/README.md (hunk 13 above)
Comment (marsishandsome) [reply mapreduce/README.md:29]: `-ts` is optional?

File: mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/example/TiDBMapreduceDemo.java (hunk 3 above)
Comment (mengxin9014) [reply mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/example/TiDBMapreduceDemo.java]: done

File: mapreduce/mapreduce-base/src/main/test/java/io/tidb/bigdata/mapreduce/tidb/MapReduceTest.java (hunk 4 above)
Comment (mengxin9014) [reply mapreduce/mapreduce-base/src/main/test/java/io/tidb/bigdata/mapreduce/tidb/MapReduceTest.java]: done

File: mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/MapreduceCmd.java (hunk 5 above)
Comment (mengxin9014) [reply mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/MapreduceCmd.java]: done

File: mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBConfiguration.java (hunk 7 above)
Comment (mengxin9014) [reply mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBConfiguration.java:81]: if field param is null, the field list in conf will be updated to all fields.

File: mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBConfiguration.java (hunk 8 above)
Comment (mengxin9014) [reply mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBConfiguration.java]: done

File: mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBInputSplit.java (hunk 9 above)
Comment (mengxin9014) [reply mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBInputSplit.java]: done

File: mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBInputSplit.java (hunk 10 above)
Comment (mengxin9014) [reply mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBInputSplit.java]: done

File: mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBResultSet.java (hunk 11 above)
Comment (mengxin9014) [reply mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBResultSet.java:440]: done

File: mapreduce/README.md (hunk 12 above)
Comment (mengxin9014) [reply mapreduce/README.md]: done

File: mapreduce/README.md (hunk 14 above)
Comment (mengxin9014) [reply mapreduce/README.md:118]: already add create table and insert record sql

File: mapreduce/README.md (hunk 13 above)
Comment (mengxin9014) [reply mapreduce/README.md:28]: add param specification and example

File: mapreduce/README.md [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+ -f   ${FIELD1} \
+ -f   ${FIELD2} \
+ -t   ${TABLE_NAME}
+ -ts  ${SNAPSHOT_TIMESTAMP} \
+ -l   ${LIMIT} 
+```
+
+### create table and insert record
+
+you could use mysql client to create table and insert record:
```
Comment (marsishandsome) [inline mapreduce/README.md:34]:
```suggestion
You could use mysql client to create the table and insert the record:
```

**Review by marsishandsome — APPROVED** [review]
others LGTM

File: mapreduce/README.md (hunk 15 above)
Comment (mengxin9014) [reply mapreduce/README.md]: done

**Comment by marsishandsome** [comment]
@sunxiaoguang PTAL

**Comment by sunxiaoguang** [comment]
> @sunxiaoguang PTAL

Sure, I will take a look later on.

**Review by sunxiaoguang — APPROVED** [review]
LGTM

**Comment by sunxiaoguang** [comment]
I made some small changes to fix build after merge from master and typo in class name. PTAL @mengxin9014 @marsishandsome

**Comment by mengxin9014** [comment]
LGTM

**Pull request merged**